rust/headless-client/src/windows.rs

```diff
@@ -0,0 +1,486 @@
+                ServiceControlHandlerResult::NoError
+            }
+            ServiceControl::UserEvent(_) => ServiceControlHandlerResult::NoError,
+            ServiceControl::Continue
+            | ServiceControl::NetBindAdd
+            | ServiceControl::NetBindDisable
+            | ServiceControl::NetBindEnable
+            | ServiceControl::NetBindRemove
+            | ServiceControl::ParamChange
+            | ServiceControl::Pause
+            | ServiceControl::Preshutdown
+            | ServiceControl::Shutdown
+            | ServiceControl::HardwareProfileChange(_)
+            | ServiceControl::PowerEvent(_)
+            | ServiceControl::SessionChange(_)
+            | ServiceControl::TimeChange
+            | ServiceControl::TriggerEvent => ServiceControlHandlerResult::NotImplemented,
+            _ => ServiceControlHandlerResult::NotImplemented,
+        }
+    };
+
+    // Fixes <https://github.com/firezone/firezone/issues/4899>,
+    // DNS rules persisting after reboot
+    connlib_shared::deactivate_dns_control().ok();
+
+    // Tell Windows that we're running (equivalent to sd_notify in systemd)
+    let status_handle = service_control_handler::register(SERVICE_NAME, event_handler)?;
+    status_handle.set_service_status(ServiceStatus {
+        service_type: SERVICE_TYPE,
+        current_state: ServiceState::Running,
+        controls_accepted: ServiceControlAccept::STOP,
+        exit_code: ServiceExitCode::Win32(0),
+        checkpoint: 0,
+        wait_hint: Duration::default(),
+        process_id: None,
+    })?;
+
+    let mut ipc_service = pin!(ipc_listen(CliIpcService::default()));
+    let result = rt.block_on(async {
+        std::future::poll_fn(|cx| {
+            match shutdown_rx.poll_recv(cx) {
+                Poll::Ready(Some(())) => {
+                    tracing::info!("Got shutdown signal");
+                    return Poll::Ready(Ok(()));
+                }
+                Poll::Ready(None) => {
+                    return Poll::Ready(Err(anyhow!(
+                        "shutdown channel unexpectedly dropped, shutting down"
+                    )))
+                }
+                Poll::Pending => {}
+            }
+
+            match ipc_service.as_mut().poll(cx) {
+                Poll::Ready(Ok(())) => {
+                    return Poll::Ready(Err(anyhow!("Impossible, ipc_listen can't return Ok")))
+                }
+                Poll::Ready(Err(error)) => {
+                    return Poll::Ready(Err(error.context("ipc_listen failed")))
+                }
+                Poll::Pending => {}
+            }
+
+            Poll::Pending
+        })
+        .await
+    });
+
+    // Tell Windows that we're stopping
+    status_handle.set_service_status(ServiceStatus {
+        service_type: SERVICE_TYPE,
+        current_state: ServiceState::Stopped,
+        controls_accepted: ServiceControlAccept::empty(),
+        exit_code: ServiceExitCode::Win32(0),
```

Comment:
According to https://stackoverflow.com/a/4197271, we should set this to a non-zero value on failure.

Reply:
Oh, _that_ exit code. Thanks, Microsoft. Fixed it.